### ISSUE TITLE
FreeBSD needs to get tests running

### ISSUE THREAD

**Issue by josteink** [dotnet/coreclr#1010]
To have any real confidence in the FreeBSD builds, we need to be able to run the associated coreclr and corefx tests.

This has been hanging in the air for quite some time, and I'm not sure we're making any progress. Therefore I'm creating an issue for better being able to track it.

Soooo with that said...

Getting the coreclr and corefx test-suite running on FreeBSD will probably involve quite a bit of sub-tasks. In that regard this task can or may be regarded as a super-task used to organize those tasks.

First sub-task is looking into getting the [coreclr tests](https://github.com/dotnet/coreclr/tree/master/tests) run. How do we get those running at all? Can they currently be run on other platforms than Windows? Is it documented how to get them running at all?

As far as I can tell, the current advisory WRT corefx and tests is that it's not all running on Unix. [Some attempts to get it running on Linux](https://github.com/dotnet/corefx/wiki/Building-On-Unix) has been made and confirmed building at least (according to @milizhang).

These instructions pre-suppose at least a few things:

* a trunk/master CI-build of mono, delivered via a Ubuntu-PPA.
* having corefx binaries to work with.

After that we'll also need xunit to run the tests.

Since we don't have a mono PPA to rely on, I guess that means we need to get it built ourselves, *and* need to document that as well. I think this will account for two sub-tasks.

For getting the corefx binaries, @MattWhilden has given us this:

> For running CoreFX stuff you'll need access to a windows machine to build binaries. If you don't have access, @ellismg has offered to be on point for jamming binaries somewhere. Let us know.

The other option is trying to build them ourselves anyway, using the [`xplat`-branch of msbuild](https://github.com/Microsoft/msbuild/tree/xplat).

@MattWhilden also thinks getting xunit running on FreeBSD might involve challenges. Based on the [test-runner script](https://github.com/dotnet/corefx/blob/master/run-test.sh), I'll assume we're talking about [this particular xunit](https://github.com/xunit/xunit/).

If so, that's mostly .NET-code itself, so it should be fairly self-validating? Also it is fetched via Nuget, so it looks like we won't have to build it ourselves. I guess we'll find out which parts Matt thinks will be challenging as we get along.

Once we have xunit-running, we'll probably have failing tests. If we get to that point, I suggest we close this issue and create new issues for ongoing test-failures.

Sounds good? Any opinions?

cc: @ghuntley @janhenke @ajensenwaud

**Comment by josteink** [comment]
> First sub-task is looking into getting the coreclr tests run. How do we get those running at all? Can they currently be run on other platforms than Windows? Is it documented how to get them running at all?

So after skimming the existing documentation slightly, I've found [this](https://github.com/dotnet/coreclr/wiki/Validating-changes):

> It may be valuable to use CoreFX tests to validate your changes to CoreCLR or mscorlib.
>
> ...
> 
> To run tests, follow the procedure for running tests in CoreFX.

As far as I can tell, those procedures are found in [the wiki](https://github.com/dotnet/corefx/wiki/Developer-Guide#tests), and they also refer to the Linux-build instructions referenced above.

So I guess there's no coreclr-tests to run, unless [buildtests.cmd](https://github.com/dotnet/coreclr/blob/master/tests/buildtest.cmd) and [runtests.cmd](https://github.com/dotnet/coreclr/blob/master/tests/runtest.cmd) is ported to Unix first.

I think this is a worthy sub-issue, also with regard to Linux and OSX support, not just FreeBSD, so I'll go ahead and create a different sub-issue for that.

**Comment by josteink** [comment]
(Parts) of this issue is preceeded by issue https://github.com/dotnet/coreclr/issues/1011 .

**Comment by jasonwilliams200OK** [comment]
There is this tool https://github.com/gajop/batch2bash. I am not sure how efficient it is in transpiling batch to bash script.
Looking into [`buildtests.cmd`](https://github.com/dotnet/coreclr/blob/master/tests/buildtest.cmd), it seems like it is calling powershell script `probe-win.ps1`, which would also require separate conversion.
@akoeplinger, by any chance, have you used the aforementioned conversion tool? Any comment on its authenticity?

Would be perfect if we have counterpart `.sh` for each `.cmd` and `.ps1` in corefx and coreclr repos.

**Comment by akoeplinger** [comment]
@jasonwilliams200OK nope, haven't used that tool yet, but the .cmd looks fairly simple so I guess the tool would work just fine. At least in corefx most of the logic happens inside MSBuild targets anyway (not sure if coreclr is the same).

**Comment by janhenke** [comment]
@josteink Sounds like a good plan.

RE Mono. So far I have the impression that the FreeBSD ports are pretty up to date. I did not see Mono 4.0 yet as an update, but I hope that comes soon.

**Comment by josteink** [comment]
It seems that @ajensenwaud is already on the task of getting a trunk/tarball based build of mono running, and already now we have 4.0.1 on the interroute host. (Wow!) Good job Anders!

**Comment by ajensenwaud** [comment]
No worries, I managed to patch the Ports makefile to 4.0.1.28, which has now been committed. This brings us closer to getting Roslyn ported, however, there are still some issues in Mono 4.0, which have been fixed in 4.1 (the development version / head). 4.1/head breaks on FreeBSD, however, so I can't get Roslyn to compile in its entirety.

What specific issues are you experiencing when running the tests?

**Comment by akoeplinger** [comment]
I think the first step is running the PAL tests as part of CI. The Linux build also only runs those right now: https://github.com/dotnet/dotnet-ci/blob/33975b34088953e9df63d431a46e2eec00c3c3d0/jobs/dotnet_coreclr_linux_release/config.xml#L47

**Comment by josteink** [comment]
> What specific issues are you experiencing when running the tests?

@ajensenwaud For now I think my biggest and most major concern is that there seems to be no (documented) way of running the CoreCLR test-suite on Unixes. I've asked around for weeks without getting any answers on this matter.

Which brings us to issue https://github.com/dotnet/coreclr/issues/1011

@akoeplinger If that's all it takes to get the tests running, it's completely bonkers how nobody has bothered to document this anywhere yet.

I'm doing a rebuild with latest sources now to give it a try. Thanks for someone (finally!) answering my question. Very curious about the results :)

**Comment by jasonwilliams200OK** [comment]
@josteink, @akoeplinger (sorry for not being able to test myself), would you please check whether these tests pass on Linux / FreeBSD: https://github.com/dotnet/coreclr/pull/1041?

**Comment by jasonwilliams200OK** [comment]
@josteink, for the first test failing, it seems like there is a note specifically for FreeBSD: https://github.com/dotnet/coreclr/blob/7a94cd137e6f737b1663706fd98c4c6b22cd2667/src/pal/tests/palsuite/threading/SleepEx/test2/test2.c#L50. Perhaps that *workaround* is obsolete?

**Comment by jasonwilliams200OK** [comment]
&lt; Just realized that the issue number is 1010! :100: &gt;

**Comment by josteink** [comment]
@jasonwilliams200OK Good observation. That workaround seems to be in all the failing tests. I'm removing it and re-running the tests.

**Comment by josteink** [comment]
I've disabled the workaround and I'm pretty sure the workaround should no longer be needed... That said the tests are still failing.

But looking at the output, it's timing based, and we can roughly say the tests fails because of deltas averagely about 10ms outside tolerance-values.

````
..............Expected thread to sleep for 1000 ms (and get interrupted).
Average delta: 69 ms,  acceptable delta: 50

FAILED: threading/SleepEx/test2/paltest_sleepex_test2. Exit code: 1

.Expected thread to wait for 2000 ms (and get interrupted).
Thread waited for 2117 ms! (Acceptable delta: 100)

FAILED: threading/WaitForSingleObject/WFSOExMutexTest/paltest_waitforsingleobject_wfsoexmutextest. Exit code: 1

.Expected thread to wait for 2000 ms (and get interrupted).
Thread waited for 2134 ms! (Acceptable delta: 100)

FAILED: threading/WaitForSingleObject/WFSOExSemaphoreTest/paltest_waitforsingleobject_wfsoexsemaphoretest. Exit code: 1

.Expected thread to wait for 2000 ms (and get interrupted).
Thread waited for 2127 ms! (Acceptable delta: 100)

FAILED: threading/WaitForSingleObject/WFSOExThreadTest/paltest_waitforsingleobject_wfsoexthreadtest. Exit code: 1
````

As you see repsectively 19, 17, 34, 27 milliseconds outside acceptable deltas. I ran another test-run where only 3 of those tests fail.

This sort of error can easily be attributed to the specific server they run on. On another server these tests might pass just fine.

I'm not *happy* about it though. Anyone have an opinion on what a canonical fix actually is here? The threading code comes from the platform and our ability to fine-tune it is limited. Should we adjust the deltas instead?

I'm open to anything.

**Comment by akoeplinger** [comment]
I think this comment in the file provides some interesting tidbits:

> /* We need to keep in mind that BSD has a timer resolution of 10ms, so 
    we need to adjust our delta to keep that in mind. Besides we need some 
    tolerance to account for different scheduling strategies, heavy load 
    scenarios, etc */ 
   const DWORD AcceptableDelta = 50; 

So it seems that this might just be an effect of the lower timer resolution. Increasing the threshold on FreeBSD sounds like a good compromise to me.

**Comment by josteink** [comment]
Trying to stay true to may "original" mission for this issue (master issue to track all issues related to getting all tests running on FreeBSD), I created a separate sub-issue for the failing tests.

Feel free to weight in there.

**Comment by ghuntley** [comment]
Related issue #1221

**Comment by francoisjacques** [comment]
FreeBSDers, has anyone looked at Rotor package to see how it was done to make the test suite running? 

Rotor is basically the same as this repo - although some parts are in Rotor but not here for some obscure reason (... networking?!?) The tests used to work just fine on BSD, although I understand the source code has probably went through several waves of changes between Rotor 1 and 2 and today...

Bottom line, the PAL suite ran fine on FreeBSD, back then.

**Comment by josteink** [comment]
Maybe we've gotten lost in the issue-tracking here now?

The CI-setup for FreeBSD should now include PAL tests:


As far as I remember *and* looking at the latest CI console-outputs, we can clearly see the **PAL** tests are getting run:

http://dotnet-ci.cloudapp.net/job/dotnet_coreclr_freebsd_debug/lastSuccessfulBuild/console

**Comment by josteink** [comment]
Maybe we've gotten lost in the issue-tracking here now?

The [CI-setup for FreeBSD](https://github.com/dotnet/dotnet-ci/blob/master/jobs/dotnet_coreclr_freebsd_debug/config.xml) should include **PAL** tests. Looking at the latest CI console-outputs, we can clearly see the [PAL tests are getting run](http://dotnet-ci.cloudapp.net/job/dotnet_coreclr_freebsd_debug/lastSuccessfulBuild/console
).

So PAL tests are fine.

This issue was actually meant to "all" dotnet/core* tests, including those for corefx, and I don't think we have those running yet, as the corefx porting effort has currently stalled a little.

Maybe it would be more helpful to limit this issue to coreclr itself, and leave the corefx tests to corefx? Or am I forgetting something?

I see the Windows-build runs a `runtests.cmd` which we still haven't ported to Unixy OSes, but that is tracked in [a separate issue](https://github.com/dotnet/coreclr/issues/1011).

With PAL-tests running, maybe it would be helpful to close this issue, and focus on the Unix-general one? Or do we still rely on fixing something from the corex-parts of the world before we're fully good to go?

**Comment by francoisjacques** [comment]
Seems reasonable to keep issues scoped by repo - it's nearly impossible to track issues covering multiple repos on github! 

Seems also reasonable to close this issue as runtests.cmd already has one issue for it...

**Comment by josteink** [comment]
Agreed. Closing this issue as fixed.

**Comment by francoisjacques** [comment]
(having another look) I realize now that the BVT was not checked-in... (or was renamed?)

Therefore the script [rrun](https://github.com/lewischeng-ms/sscli/blob/adbd6bbad0816f65650a66e1701112a307dc4b09/tests/rrun.pl) that is used to run such base tests and the BCL tests is not present and has to be rewritten, somehow?

**Comment by josteink** [comment]
Sounds like something which affects all Unix platforms equally and like material for a new issue?
